### ISSUE TITLE
Eclipse Che with Let's Encrypt Certificate

### ISSUE THREAD

**Issue by AussieScorpio** [eclipse/che#22546]
### Summary

I have Let's Encrypt as an issuer and it correctly creates certificates when I have an Ingress with secretName: letsencrypt-production. Routing works when I have a valid Ingress with ingressClassName: nginx.

My CheCluster is;
```
apiVersion: org.eclipse.che/v2
kind: CheCluster
metadata:
  name: eclipse-che
  namespace: eclipse-che
spec:
  components: {}
  containerRegistry: {}
  devEnvironments: {}
  gitServices: {}
  networking:
    domain: eclipse.my.domain
    ingressClassName: nginx
    tlsSecretName: letsencrypt-production 
    auth:
      identityProviderURL: https://sso.my.domain/realms/eclipse-che/
      oAuthClientName: eclipse-che
      oAuthSecret: pN9HFPh5kj5XfQBqIcfpnI9fO5zglTLE
```

When I attempt to navigate to eclipse.my.domain, I'm getting a warning about a bad certificate. When I inspect the certificate, I see that it's a certificate created by Eclipse Che. The Ingress has the correct secretName, so tlsSecretName: letsencrypt-production is being correctly picked up, yet I'm getting an invalid Eclipse Che certificate rather than the Let's Encrypt certificate.

How is this meant to work? How do I get Eclipse Che to just let Let's Encrypt create the certificate?

**Comment by tolusha** [comment]
Hello, @AussieScorpio 
Basically you have to create two secrets: che-tls and self-signed-certificate. 
I hope it will help you [1].

[1] https://eclipse.dev/che/docs/stable/administration-guide/configuring-che-with-self-signed-certificate

**Comment by AussieScorpio** [comment]
> Hello, @AussieScorpio Basically you have to create two secrets: che-tls and self-signed-certificate. I hope it will help you [1].
> 
> [1] https://eclipse.dev/che/docs/stable/administration-guide/configuring-che-with-self-signed-certificate

This documentation is referring to a self-signed certificate. I want to use a Let's Encrypt certificate. My cluster creates these automatically for a valid Ingress. Am I missing something?

**Comment by tolusha** [comment]
The idea is to have those 2 secrets in eclipse-che namespace BEFORE deploying Eclipse Che.
- self-signed-certificate
- che-tls (or other name defined in spec.networking.tlsSecretName)

**Comment by AussieScorpio** [comment]
> The idea is to have those 2 secrets in eclipse-che namespace BEFORE deploying Eclipse Che. They can be generated by different ways:
> 
> * self-signed-certificate
> * che-tls (or other name defined in spec.networking.tlsSecretName)

It's not clear to me 1) why we need two secrets, and 2) why, when the Ingress has been specified, that it's not honoured.

The documentation might be clear to someone who knows Eclipse Che intimately, but is not very useful for helping someone new to Eclipse Che to get up to speed. A series of prescriptive steps as detailed in the link, but without any context as to why we're performing these steps, or how the resources are referenced or used makes it very difficult to troubleshoot when something doesn't work as expected.

Is there any documentation on what resources Eclipse Che uses, and how it all hangs together; including how the installer works?

**Comment by tolusha** [comment]
I agree the documentation might not be clear. 
I would be nice if someone else contribute into Eclipse Che project to improve the documentation.

Let me clarify:
1. Eclipse Che operator creates the ingress [1] using the secret specified in the CheCluster CR (`spec.networking.tlsSecretName`)
2. If Secret doesn't exist then certificates will be generated and secret created in the eclipse-che namespace.

Could you tell me what does Secret `letsencrypt-production` contain? 

[1] https://github.com/eclipse-che/che-operator/blob/c927cac2bf7129511d3b1d57710ca7e69edb3bbc/pkg/deploy/ingress.go#L65

**Comment by l0rd** [comment]
@tolusha I don''t think that `self-signed-certificate` certificate is needed. Let's encrypt CA is known and trusted by browsers so I don't think we need to add the CA cert.

[The documentation of CheCluster `.spec.networking.tlsSecretName` doesn't mention that](https://doc.crds.dev/github.com/eclipse-che/che-operator/org.eclipse.che/CheCluster/v2@7.72.0#spec-networking-tlsSecretName):

```
The name of the secret used to set up Ingress TLS termination. 
If the field is an empty string, the default cluster certificate is 
used. The secret must have a app.kubernetes.io/part-of=che.eclipse.org label.
```

**Comment by AussieScorpio** [comment]
> I agree the documentation might not be clear. I would be nice if someone else contribute into Eclipse Che project to improve the documentation.
> 
> Let me clarify:
> 
> 1. Eclipse Che operator creates the ingress [1] using the secret specified in the CheCluster CR (`spec.networking.tlsSecretName`)
> 2. If Secret doesn't exist then certificates will be generated and secret created in the eclipse-che namespace.
> 
> Could you tell me what does Secret `letsencrypt-production` contain?
> 
> [1] https://github.com/eclipse-che/che-operator/blob/c927cac2bf7129511d3b1d57710ca7e69edb3bbc/pkg/deploy/ingress.go#L65

If I could make one comment about a lot of the documentation; much of it is instructional rather than documentary, and much of it provides detail without context. For someone coming in cold to something as complex as Eclipse Che; it's important to be able to gain a "mental model" to provide context to documentation. A high level introduction to how it hangs together and essential concepts (as though you're explaining it to a 10 year old) coupled with references back to the introduction in the documentation to provide context would be invaluable; and give a far greater return on the time and energy that goes into documentation.

**Comment by AussieScorpio** [comment]
> @tolusha I don''t think that `self-signed-certificate` certificate is needed. Let's encrypt CA is known and trusted by browsers so I don't think we need to add the CA cert.
> 
> [The documentation of CheCluster `.spec.networking.tlsSecretName` doesn't mention that](https://doc.crds.dev/github.com/eclipse-che/che-operator/org.eclipse.che/CheCluster/v2@7.72.0#spec-networking-tlsSecretName):
> 
> ```
> The name of the secret used to set up Ingress TLS termination. 
> If the field is an empty string, the default cluster certificate is 
> used. The secret must have a app.kubernetes.io/part-of=che.eclipse.org label.
> ```

I've had another crack at it; this time providing annotations to appear in the Ingress in the hope that the certificate will be created by Let's Encrypt. It still doesn't work.

Installing Che on a cluster that already has a certificate issuer must be the most common thing in the world. I can't believe that everyone in a production environment is using self-signed certificates.

How are others doing this?

```
apiVersion: org.eclipse.che/v2
kind: CheCluster
metadata:
  name: eclipse-che
  namespace: eclipse-che
spec:
  components: {}
  containerRegistry: {}
  devEnvironments: {}
  gitServices: {}
  networking:
    domain: eclipse.genesys.works
    annotations:
      acme.cert-manager.io/http01-edit-in-place: "true"
      cert-manager.io/cluster-issuer: letsencrypt-production
      nginx.ingress.kubernetes.io/ssl-redirect: "true"
    ingressClassName: nginx
    tlsSecretName: letsencrypt-production
    auth:
      identityProviderURL: https://sso.ethnique.biz/realms/eclipse-che/
      oAuthClientName: eclipse-che
      oAuthSecret: pN9HFPh5kj5XfQBqIcfLnI9fO5zglTLE
```

**Comment by tolusha** [comment]
I apologize if I misunderstood you, maybe we talk about different things.
I don't see any issues if certificate issuer already exists on the cluster.

For instance:
```
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: che-tls
  namespace: eclipse-che
spec:
  ...
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: che-tls
  namespace: eclipse-che
spec:
  issuerRef:
    kind: Issuer
    name: che-tls
    group: cert-manager.io
  secretName: che-tls
  ...
 ```
 
 The idea is to create `che-tls` secret and label it appropriately: `oc label secrets -n eclipse-che che-tls app.kubernetes.io/part-of=che.eclipse.org`. After that we can deploy Eclipse che with chectl without specifying anything in networking.annotations and networking.tlsSecretName fields.
 
 @l0rd 
 You are right, we don't need Secret `self-signed-certificate` in case of Let's encrypt.

**Comment by AussieScorpio** [comment]
> I apologize if I misunderstood you, maybe we talk about different things. I don't see any issues if certificate issuer already exists on the cluster.
> 
> For instance:
> 
> ```
> apiVersion: cert-manager.io/v1
> kind: Issuer
> metadata:
>   name: che-tls
>   namespace: eclipse-che
> spec:
>   ...
> ---
> apiVersion: cert-manager.io/v1
> kind: Certificate
> metadata:
>   name: che-tls
>   namespace: eclipse-che
> spec:
>   issuerRef:
>     kind: Issuer
>     name: che-tls
>     group: cert-manager.io
>   secretName: che-tls
>   ...
> ```
> 
> The idea is to create `che-tls` secret and label it appropriately: `oc label secrets -n eclipse-che che-tls app.kubernetes.io/part-of=che.eclipse.org`. After that we can deploy Eclipse che with chectl without specifying anything in networking.annotations and networking.tlsSecretName fields.
> 
> @l0rd You are right, we don't need Secret `self-signed-certificate` in case of Let's encrypt.

I'm still missing a piece of the puzzle somewhere.

So is it the case that Eclipse Che cannot use my clusterissuer? If that's the case, do I need to copy the spec from the clusterissuer (which has the Let's Encrypt config) to an issuer resource and then manually create the certificate? Can I manually create the certificate using the clusterissuer? And if I'm manually creating the certificate, as i understand, the certificate won't be issued and the secret created until such time as the ingress has been created in order for the HTTP-01 challenge to succeed. If I'm also having to manually create this, then what's the point of the installer? Why is it not possible to define this once in the checluster resource?

**Comment by tolusha** [comment]
> Can I manually create the certificate using the clusterissuer?

Sure, you can.

> And if I'm manually creating the certificate, as i understand, the certificate won't be issued and the secret created until such time as the ingress has been created in order for the HTTP-01 challenge to succeed

Secret will be created independently of ingress. I don't see why it won't happened. 
Just to mention, that Secret must be labeled respectively `oc label secrets -n eclipse-che che-tls app.kubernetes.io/part-of=che.eclipse.org`

**Comment by AussieScorpio** [comment]
> > Can I manually create the certificate using the clusterissuer?
> 
> Sure, you can.
> 
> > And if I'm manually creating the certificate, as i understand, the certificate won't be issued and the secret created until such time as the ingress has been created in order for the HTTP-01 challenge to succeed
> 
> Secret will be created independently of ingress. I don't see why it won't happened. Just to mention, that Secret must be labeled respectively `oc label secrets -n eclipse-che che-tls app.kubernetes.io/part-of=che.eclipse.org`

How will the challenge work?

**Comment by AussieScorpio** [comment]
Ok. I seem to be able to get Eclipse Che deployed with a Let's Encrypt certificate. Here's what I did;

Using this as a reference; https://artifacthub.io/packages/helm/eclipse-che/eclipse-che

>helm repo add eclipse-che https://eclipse-che.github.io/che-operator/charts

>kubectl apply -f init

>kubectl label secrets -n eclipse-che che-tls app.kubernetes.io/part-of=che.eclipse.org

>helm install eclipse-che eclipse-che/eclipse-che --version 7.74.0 --namespace eclipse-che --create-namespace

>kubectl apply -f org_v2_checluster.yaml -n eclipse-che

The files referenced are;
init/01-namespace.yaml
`apiVersion: v1
kind: Namespace
metadata:
  name: eclipse-che
`

init/certificate.yaml
`apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  annotations:
    acme.cert-manager.io/http01-override-ingress-name: che
    cert-manager.io/issue-temporary-certificate: "true"
  name: che-tls
  namespace: eclipse-che
spec:
  dnsNames:
  - eclipse.mydomain.com
  issuerRef:
    group: cert-manager.io
    kind: ClusterIssuer
    name: letsencrypt-production
  secretName: che-tls
  usages:
  - digital signature
  - key encipherment`

org_v2_checluster.yaml
`apiVersion: org.eclipse.che/v2
kind: CheCluster
metadata:
  name: eclipse-che
  namespace: eclipse-che
spec:
  components: {}
  containerRegistry: {}
  devEnvironments: {}
  gitServices: {}
  networking:
    domain: eclipse.mydomain.com
    annotations:
      acme.cert-manager.io/http01-edit-in-place: "true"
      cert-manager.io/cluster-issuer: letsencrypt-production
      nginx.ingress.kubernetes.io/ssl-redirect: "true"
    ingressClassName: nginx
    tlsSecretName: che-tls
    auth:
      identityProviderURL: https://sso.mydomain.com/realms/eclipse-che/
      oAuthClientName: eclipse-che
      oAuthSecret: pN9HFPh5kj5XfQBqIcfLoI9fO5zglTLE`

This all assumes a functional cluster with nginx-ingress and cert-manager installed with an instance of Keycloak (in my case on the same cluster). A clusterissuer exists for Let's Encrypt with a name of letsencrypt-production.

Some have suggested having a label on the clusterissuer where you have more than one `client/org.eclipse.che=true`. In my case, having this label prevented creation of certificates. Once I removed the label it worked.

The annotation `acme.cert-manager.io/http01-override-ingress-name` for the certificate turned out to be critical as Eclipse Che breaks convention and refers to itself as simply 'che' for the ingress.

It may be that some of this is unnecessary, and maybe there's a better way (it seems to me that there should be a way to define the clusterissuer to use in the checluster and be done with it).

On to the next challenge; figuring out why I can login and get to the dashboard, but get;

_Failed to fetch available workspaces, reason: Failed to fetch the list of devWorkspaces. Unable to list devworkspaces: Unauthorized

Danger alert:Failed to fetch the user profile data. Unable to get user profile data: Unauthorized_

I'll open a separate post for this.
